File: android/src/main/java/com/octopuscommunity/octopusreactnativesdk/OctopusSDKInitializer.kt
package com.octopuscommunity.octopusreactnativesdk

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.octopuscommunity.sdk.ApiServer
import com.octopuscommunity.sdk.OctopusSDK
import com.octopuscommunity.sdk.domain.model.ConnectionMode
import android.graphics.Color

class OctopusSDKInitializer {

  fun initialize(context: ReactApplicationContext, options: ReadableMap, promise: Promise): Boolean {
    val apiKey = options.getString("apiKey")
    if (apiKey == null) {
      promise.reject("INITIALIZE_ERROR", "Missing API key")
      return false
    }

    try {
      val connectionMode = parseConnectionMode(options)
      // Parity wave — lifecycle
      val deepLinksBasePaths = parseDeepLinksBasePaths(options)
      val apiServer = parseApiServer(options)

      // Store theme configuration for later use in UI
      val themeConfig = parseThemeConfig(options)
      OctopusThemeManager.setThemeConfig(themeConfig)

      // Store UI configuration separately
      val uiConfiguration = parseUIConfiguration(options)
      OctopusUIConfigurationManager.setUIConfiguration(uiConfiguration)

      val topAppBarConfig = parseTopAppBarConfig(options)
      OctopusTopAppBarManager.setConfig(topAppBarConfig)

      OctopusSDK.initialize(
        context = context,
        apiKey = apiKey,
        connectionMode = connectionMode,
        deepLinksBasePaths = deepLinksBasePaths,
        apiServer = apiServer
      )
      promise.resolve(null)
      return true
    } catch (e: InvalidConnectionModeException) {
      promise.reject("INITIALIZE_ERROR", e.message, e)
      return false
    } catch (e: Exception) {
      // Parity wave — lifecycle: was a fixed string, discarding e.message. That message is now
      // the only place an invalid `apiServer` host (thrown from the native `ApiServer`
      // constructor — see parseApiServer below) surfaces on Android, matching iOS forwarding
      // `error.localizedDescription`.
      promise.reject("INITIALIZE_ERROR", e.message ?: "Failed to initialize Octopus SDK", e)
      return false
    }
  }

  fun parseThemeConfig(options: ReadableMap): OctopusThemeConfig? {

    val themeMap = options.getMap("theme")
    val colorScheme = options.getString("colorScheme") // Get colorScheme from React Native

    var primaryColor: String? = null
    var primaryLowContrastColor: String? = null
    var primaryHighContrastColor: String? = null
    var onPrimaryColor: String? = null
    var linkColor: String? = null
    var backgroundColor: String? = null
    var logoSource: ReadableMap? = null
    var fontsConfig: OctopusFontsConfig? = null
    var lightSet: OctopusModeColors? = null
    var darkSet: OctopusModeColors? = null

    // Parse colors from theme if available
    themeMap?.let { theme ->
      val colorsMap = theme.getMap("colors")
      colorsMap?.let { colors ->
        // Check if this is a dual-mode theme (has light and dark properties)
        val lightColors = colors.getMap("light")
        val darkColors = colors.getMap("dark")

        if (lightColors != null && darkColors != null) {
          // Dual-mode theme. Both raw sets are kept on the config so the render can re-select
          // for the mode in effect then (OctopusThemeConfig.resolvedFor); the flat slots below
          // are the selection for the scheme known now, kept for the first render.
          lightSet = modeColors(lightColors)
          darkSet = modeColors(darkColors)
          val selectedColors = if (colorScheme == "dark") darkColors else lightColors
          val otherColors = if (colorScheme == "dark") lightColors else darkColors
          primaryColor = colorFrom(selectedColors, "primary")
          primaryLowContrastColor = colorFrom(selectedColors, "primaryLowContrast")
          primaryHighContrastColor = colorFrom(selectedColors, "primaryHighContrast")
          onPrimaryColor = colorFrom(selectedColors, "onPrimary")
          // `link` and `background` are optional in a way `primary` is not: a value given
          // for one mode only applies to both, matching the single-value contract iOS
          // implements through its adaptive colors.
          linkColor = colorFrom(selectedColors, "link") ?: colorFrom(otherColors, "link")
          backgroundColor = colorFrom(selectedColors, "background") ?: colorFrom(otherColors, "background")
        } else {
          // Single-mode theme (backward compatibility)
          primaryColor = colorFrom(colors, "primary")
          primaryLowContrastColor = colorFrom(colors, "primaryLowContrast")
          primaryHighContrastColor = colorFrom(colors, "primaryHighContrast")
          onPrimaryColor = colorFrom(colors, "onPrimary")
          linkColor = colorFrom(colors, "link")
          backgroundColor = colorFrom(colors, "background")
        }
      }

      // Handle logo from theme
      val logoMap = theme.getMap("logo")
      logoMap?.let { logo ->
        val imageSource = logo.getMap("image")
        if (imageSource != null) {
          logoSource = imageSource
        }
      }

      // Handle fonts from theme - use pre-processed configuration from TypeScript layer
      val fontsMap = theme.getMap("fonts")
      fontsMap?.let { fonts ->
        val parsedConfig = fonts.getMap("parsedConfig")
        if (parsedConfig != null) {
          fontsConfig = parsePreProcessedFontsConfig(parsedConfig)
        }
      }
    }

    // Handle logo at root level (for backward compatibility)
    if (logoSource == null) {
      val rootLogoMap = options.getMap("logo")
      rootLogoMap?.let { logo ->
        val imageSource = logo.getMap("image")
        if (imageSource != null) {
          logoSource = imageSource
        }
      }
    }

    // Only create theme config if we have at least one customization. `link` and
    // `background` count: a theme carrying only one of them must not be dropped.
    if (primaryColor != null || linkColor != null || backgroundColor != null ||
      logoSource != null || colorScheme != null || fontsConfig != null
    ) {
      return OctopusThemeConfig(
        primaryColor = primaryColor,
        primaryLowContrastColor = primaryLowContrastColor,
        primaryHighContrastColor = primaryHighContrastColor,
        onPrimaryColor = onPrimaryColor,
        linkColor = linkColor,
        backgroundColor = backgroundColor,
        logoSource = logoSource,
        colorScheme = colorScheme,
        fonts = fontsConfig,
        lightColors = lightSet,
        darkColors = darkSet
      )
    }

    return null
  }

  fun parseUIConfiguration(options: ReadableMap): OctopusUIConfiguration? {
    val uiMap = options.getMap("ui")
    var bottomContentPadding: Double? = null

    uiMap?.let { uiOptions ->
      val candidateKeys = listOf("bottomSafeAreaInset", "bottomPadding", "contentPadding", "contentPaddingBottom")
      for (key in candidateKeys) {
        if (uiOptions.hasKey(key)) {
          bottomContentPadding = try {
            uiOptions.getDouble(key).coerceAtLeast(0.0)
          } catch (e: Exception) {
            null
          }
          if (bottomContentPadding != null) {
            break
          }
        }
      }
    }

    return if (bottomContentPadding != null) {
      OctopusUIConfiguration(bottomContentPadding = bottomContentPadding)
    } else {
      null
    }
  }

  fun parseTopAppBarConfig(options: ReadableMap): OctopusTopAppBarConfig? {
    val map = options.getMap("topAppBar") ?: return null

    var titleType: String? = null
    var titleText: String? = null
    map.getMap("title")?.let { title ->
      titleType = if (title.hasKey("type")) title.getString("type") else null
      titleText = if (title.hasKey("text")) title.getString("text") else null
    }

    val centered = map.hasKey("alignment") && map.getString("alignment") == "center"
    val coloredBackground = map.hasKey("coloredBackground") && map.getBoolean("coloredBackground")

    return OctopusTopAppBarConfig(
      titleType = titleType,
      titleText = titleText,
      centered = centered,
      coloredBackground = coloredBackground
    )
  }

  private fun parsePreProcessedFontsConfig(parsedConfig: ReadableMap): OctopusFontsConfig? {
    val textStylesMap = parsedConfig.getMap("textStyles")
    val textStyles = mutableMapOf<String, OctopusTextStyleConfig>()

    // Parse pre-processed font configuration from TypeScript layer
    textStylesMap?.let { textStylesMap ->
      // `navBarItem` is deliberately NOT read here: the native iOS theme has a
      // dedicated `OctopusTheme.Fonts.navBarItem` slot, but the native Android
      // `OctopusTypography` has no counterpart (title1/title2/body1/body2/
      // caption1/caption2 only), so there is nothing to map it onto. Do NOT
      // approximate it by resizing another slot — that would change text the
      // host did not ask to change.
      val textStyleKeys = arrayOf("title1", "title2", "body1", "body2", "caption1", "caption2")

      textStyleKeys.forEach { key ->
        val textStyleMap = textStylesMap.getMap(key)
        textStyleMap?.let { style ->
          val fontType = style.getString("fontType")
          val fontSize = if (style.hasKey("fontSize")) style.getDouble("fontSize") else Double.NaN

          if (fontType != null || (!fontSize.isNaN() && fontSize > 0)) {
            textStyles[key] = OctopusTextStyleConfig(
              fontType = fontType,
              fontSize = if (fontSize.isNaN() || fontSize <= 0) null else fontSize
            )
          }
        }
      }
    }

    // Theme-wide overrides. Both are forwarded verbatim: `fontFamily` names a `res/font/`
    // resource only the render pass can look up, and `fontWeight` has already been
    // validated against the 100-900 scale by the TypeScript layer.
    val fontFamily = parsedConfig.getString("fontFamily")?.takeIf { it.isNotBlank() }
    val fontWeight = if (parsedConfig.hasKey("fontWeight") && !parsedConfig.isNull("fontWeight")) {
      parsedConfig.getDouble("fontWeight").toInt()
    } else {
      null
    }

    // Only create fonts config if we have text styles or a theme-wide override: a theme
    // made only of `fontFamily` / `fontWeight` must not be dropped.
    if (textStyles.isNotEmpty() || fontFamily != null || fontWeight != null) {
      return OctopusFontsConfig(
        textStyles = textStyles,
        fontFamily = fontFamily,
        fontWeight = fontWeight
      )
    }

    return null
  }


  private fun extractResourceNameFromUri(uri: String): String? {
    // Extract resource name from React Native image URI
    // Examples: "logo.png" -> "logo", "images/logo.png" -> "logo"
    return try {
      val fileName = uri.substringAfterLast("/")
      fileName.substringBeforeLast(".")
    } catch (e: Exception) {
      null
    }
  }

  private fun modeColors(colors: ReadableMap) = OctopusModeColors(
    primary = colorFrom(colors, "primary"),
    primaryLowContrast = colorFrom(colors, "primaryLowContrast"),
    primaryHighContrast = colorFrom(colors, "primaryHighContrast"),
    onPrimary = colorFrom(colors, "onPrimary"),
    link = colorFrom(colors, "link"),
    background = colorFrom(colors, "background")
  )

  private fun colorFrom(colors: ReadableMap, key: String): String? =
    parseColor(if (colors.hasKey(key)) colors.getString(key) else null)

  private fun parseColor(colorString: String?): String? {
    if (colorString == null) return null

    return try {
      // Validate that the color string is a valid hex color
      Color.parseColor(colorString)
      // Return the original string if parsing succeeds
      colorString
    } catch (e: IllegalArgumentException) {
      // Invalid color format - return null to skip this color
      null
    }
  }

  fun parseConnectionMode(options: ReadableMap): ConnectionMode {
    val connectionModeMap = options.getMap("connectionMode")
    return when (val connectionModeType = connectionModeMap?.getString("type")) {
      "sso" -> {
        ConnectionMode.SSO(
          appManagedFields = ProfileFieldMapper.fromReactNativeArray(
            if (connectionModeMap.hasKey("appManagedFields")) connectionModeMap.getArray("appManagedFields") else null
          )
        )
      }

      "octopus" -> {
        // Parity wave — lifecycle: this used to return null here, relying on
        // OctopusSDK.initialize()'s default parameter value for `connectionMode` — which is
        // ConnectionMode.SSO(), NOT OctopusAuth. That silently put every "octopus" mode caller
        // into SSO mode with an empty appManagedFields set. Construct the correct value
        // explicitly instead.
        ConnectionMode.OctopusAuth
      }

      else -> {
        throw InvalidConnectionModeException("Invalid connection mode type: $connectionModeType")
      }
    }
  }

  /**
   * Parity wave — lifecycle.
   *
   * Extracts the deep link base paths forwarded to native `initialize`/`switchCommunity` as
   * `deepLinksBasePaths`. Only the octopus-managed connection mode carries a `deepLink` option
   * on the TS side (`InitializeParams.connectionMode` / `SwitchCommunityParams.connectionMode`),
   * used by the native SDK to build the magic-link confirmation redirect
   * (`OctopusDeepLinks.octopusAppNavDeepLinks`). The single value maps onto the native
   * `List<String>` parameter as its sole entry; `emptyList()` when absent, matching the native
   * default.
   */
  fun parseDeepLinksBasePaths(options: ReadableMap): List<String> {
    val connectionModeMap = options.getMap("connectionMode") ?: return emptyList()
    if (connectionModeMap.getString("type") != "octopus") return emptyList()
    val deepLink = if (connectionModeMap.hasKey("deepLink")) {
      connectionModeMap.getString("deepLink")
    } else {
      null
    }
    return listOfNotNull(deepLink)
  }

  /**
   * Parity wave — lifecycle.
   *
   * Parses an optional custom server endpoint from `options.apiServer`. This is a pure
   * passthrough — validation (`ApiServer`'s host/port rules) happens natively: an invalid
   * value throws from the `ApiServer` constructor and is caught by the generic `Exception`
   * handler in [initialize] / by `OctopusReactModule.switchCommunity`.
   */
  fun parseApiServer(options: ReadableMap): ApiServer? {
    val apiServerMap = options.getMap("apiServer") ?: return null
    val host = apiServerMap.getString("host") ?: return null
    return if (apiServerMap.hasKey("port") && !apiServerMap.isNull("port")) {
      ApiServer(host = host, port = apiServerMap.getInt("port"))
    } else {
      ApiServer(host = host)
    }
  }

  private class InvalidConnectionModeException(message: String) : Exception(message)
}

File: ios/OctopusSDKInitializer.swift
import Octopus
import OctopusUI
import SwiftUI
import CoreGraphics

class OctopusSDKInitializer {
  func initialize(options: [String: Any], eventManager: OctopusEventManager) throws -> OctopusSDK {
    guard let apiKey = options["apiKey"] as? String else {
      throw InitializationError.missingAPIKey
    }

    guard let connectionModeMap = options["connectionMode"] as? [String: Any] else {
      throw InitializationError.missingConnectionMode
    }

    let connectionMode = try parseConnectionMode(from: connectionModeMap, eventManager: eventManager)
    let configuration = try parseConfiguration(from: options)
    return try OctopusSDK(apiKey: apiKey, connectionMode: connectionMode, configuration: configuration)
  }

  // Parity wave — lifecycle

  /// Builds the SDK `Configuration` from the JS `apiServer` option, mirroring the Flutter
  /// plugin's `parseConfiguration` helper. Returns the default configuration (no custom
  /// server) when `apiServer` is absent or malformed.
  func parseConfiguration(from options: [String: Any]) throws -> OctopusSDK.Configuration {
    guard let apiServerMap = options["apiServer"] as? [String: Any],
          let host = apiServerMap["host"] as? String
    else {
      return OctopusSDK.Configuration()
    }
    let port = (apiServerMap["port"] as? NSNumber)?.intValue
      ?? (apiServerMap["port"] as? Int)
      ?? 443
    let apiServer = try OctopusSDK.Configuration.ApiServer(host: host, port: port)
    return OctopusSDK.Configuration(apiServer: apiServer)
  }
  
  func parseTheme(from options: [String: Any]) -> OctopusTheme? {
    guard let themeMap = options["theme"] as? [String: Any] else {
      return nil
    }
    
    var colors: OctopusTheme.Colors?
    var fonts: OctopusTheme.Fonts?
    
    // Parse colors
    if let colorsMap = themeMap["colors"] as? [String: Any] {
      var primarySet: OctopusTheme.Colors.ColorSet?
      var onPrimary: Color?
      var link: Color?
      var background: Color?

      // Check if this is a dual-mode theme (has light and dark properties)
      if let lightColors = colorsMap["light"] as? [String: Any],
         let darkColors = colorsMap["dark"] as? [String: Any] {
        // Dual-mode theme - create adaptive colors that automatically respond to system appearance
        if let lightPrimary = lightColors["primary"] as? String,
           let darkPrimary = darkColors["primary"] as? String,
           let lightPrimaryColor = OctopusColorUtility.color(fromHex: lightPrimary),
           let darkPrimaryColor = OctopusColorUtility.color(fromHex: darkPrimary) {

          // Create adaptive colors using UIColor's dynamic color capabilities
          let adaptivePrimary = UIColor { traitCollection in
            traitCollection.userInterfaceStyle == .dark ? darkPrimaryColor : lightPrimaryColor
          }

          let lightLowContrast = (lightColors["primaryLowContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? lightPrimaryColor
          let darkLowContrast = (darkColors["primaryLowContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? darkPrimaryColor
          let adaptiveLowContrast = UIColor { traitCollection in
            traitCollection.userInterfaceStyle == .dark ? darkLowContrast : lightLowContrast
          }

          let lightHighContrast = (lightColors["primaryHighContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.black
          let darkHighContrast = (darkColors["primaryHighContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.black
          let adaptiveHighContrast = UIColor { traitCollection in
            traitCollection.userInterfaceStyle == .dark ? darkHighContrast : lightHighContrast
          }

          let lightOnPrimary = (lightColors["onPrimary"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.white
          let darkOnPrimary = (darkColors["onPrimary"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.black
          let adaptiveOnPrimary = UIColor { traitCollection in
            traitCollection.userInterfaceStyle == .dark ? darkOnPrimary : lightOnPrimary
          }

          primarySet = OctopusTheme.Colors.ColorSet(
            main: Color(uiColor: adaptivePrimary),
            lowContrast: Color(uiColor: adaptiveLowContrast),
            highContrast: Color(uiColor: adaptiveHighContrast)
          )
          onPrimary = Color(uiColor: adaptiveOnPrimary)
        } else {
          // No usable primary pair: `onPrimary` alone carries no default to derive, so it
          // is only forwarded when explicitly given.
          onPrimary = adaptiveThemeColor(light: lightColors["onPrimary"], dark: darkColors["onPrimary"])
        }

        link = adaptiveThemeColor(light: lightColors["link"], dark: darkColors["link"])
        background = adaptiveThemeColor(light: lightColors["background"], dark: darkColors["background"])
      } else {
        // Single-mode theme (backward compatibility)
        if let primary = colorsMap["primary"] as? String,
           let primaryColor = OctopusColorUtility.color(fromHex: primary) {
          let lowContrast = (colorsMap["primaryLowContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? primaryColor
          let highContrast = (colorsMap["primaryHighContrast"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.black
          let resolvedOnPrimary = (colorsMap["onPrimary"] as? String).flatMap(OctopusColorUtility.color(fromHex:)) ?? UIColor.white

          primarySet = OctopusTheme.Colors.ColorSet(
            main: Color(uiColor: primaryColor),
            lowContrast: Color(uiColor: lowContrast),
            highContrast: Color(uiColor: highContrast)
          )
          onPrimary = Color(uiColor: resolvedOnPrimary)
        } else {
          onPrimary = themeColor(colorsMap["onPrimary"])
        }

        link = themeColor(colorsMap["link"])
        background = themeColor(colorsMap["background"])
      }

      // A theme carrying only `link` or only `background` must not be dropped, so the
      // color set is built as soon as any one of the four resolved.
      if primarySet != nil || onPrimary != nil || link != nil || background != nil {
        colors = OctopusTheme.Colors(
          primarySet: primarySet,
          onPrimary: onPrimary,
          link: link,
          background: background
        )
      }
    }

    // Parse fonts
    if let fontsMap = themeMap["fonts"] as? [String: Any] {
      // Font customization will be applied through OctopusTypography
      // This is parsed here but applied separately in the UI manager
      fonts = OctopusTheme.Fonts()
    }
    
    // Note: Logo will be handled separately in UI manager due to async loading requirements
    // Only create theme if we have colors or fonts, otherwise return nil
    guard colors != nil || fonts != nil else {
      return nil
    }
    
    return OctopusTheme(
      colors: colors ?? OctopusTheme.Colors(),
      fonts: fonts ?? OctopusTheme.Fonts(),
      assets: OctopusTheme.Assets()
    )
  }
  
  /// Resolves a single hex color coming from the JS theme map.
  /// - Parameter value: the raw dictionary value; anything that is not a parseable hex
  ///                    string yields `nil` so the native default applies.
  private func themeColor(_ value: Any?) -> Color? {
    guard let hex = value as? String,
          let uiColor = OctopusColorUtility.color(fromHex: hex) else {
      return nil
    }
    return Color(uiColor: uiColor)
  }

  /// Builds a light/dark adaptive color from a dual-mode theme.
  ///
  /// A value supplied for one mode only applies to both: these colors are optional and
  /// carry no per-mode default this bridge could reference, and the Android bridge does
  /// the same for its selected color scheme.
  private func adaptiveThemeColor(light: Any?, dark: Any?) -> Color? {
    let lightColor = (light as? String).flatMap(OctopusColorUtility.color(fromHex:))
    let darkColor = (dark as? String).flatMap(OctopusColorUtility.color(fromHex:))
    guard let resolvedLight = lightColor ?? darkColor,
          let resolvedDark = darkColor ?? lightColor else {
      return nil
    }
    let adaptive = UIColor { traitCollection in
      traitCollection.userInterfaceStyle == .dark ? resolvedDark : resolvedLight
    }
    return Color(uiColor: adaptive)
  }

  func getLogoSource(from options: [String: Any]) -> [String: Any]? {
    // First try to get logo from theme
    if let themeMap = options["theme"] as? [String: Any],
       let logoMap = themeMap["logo"] as? [String: Any],
       let imageSource = logoMap["image"] as? [String: Any] {
      return imageSource
    }
    
    // Fallback: try to get logo from root level (for backward compatibility)
    if let logoMap = options["logo"] as? [String: Any],
       let imageSource = logoMap["image"] as? [String: Any] {
      return imageSource
    }
    
    return nil
  }
  
  func getFontConfiguration(from options: [String: Any]) -> [String: Any]? {
    guard let themeMap = options["theme"] as? [String: Any],
          let fontsMap = themeMap["fonts"] as? [String: Any] else {
      return nil
    }
    
    // Use pre-processed configuration from TypeScript layer
    if let parsedConfig = fontsMap["parsedConfig"] as? [String: Any] {
      return ["parsedConfig": parsedConfig]
    }
    
    return nil
  }

  func parseUIConfiguration(from options: [String: Any]) -> OctopusUIConfiguration? {
    guard let uiOptions = options["ui"] as? [String: Any] else {
      return nil
    }

    let supportedKeys = [
      "bottomSafeAreaInset",
      "bottomPadding",
      "contentPadding",
      "contentPaddingBottom",
    ]

    var bottomInset: CGFloat?

    for key in supportedKeys {
      if let value = uiOptions[key] as? NSNumber {
        bottomInset = CGFloat(truncating: value)
        break
      } else if let value = uiOptions[key] as? Double {
        bottomInset = CGFloat(value)
        break
      }
    }

    guard let finalInset = bottomInset else {
      return nil
    }

    return OctopusUIConfiguration(bottomSafeAreaInset: max(0, finalInset))
  }

  func parseTopAppBar(from options: [String: Any]) -> OctopusTopAppBarConfig? {
    guard let map = options["topAppBar"] as? [String: Any] else {
      return nil
    }

    var titleType: String? = nil
    var titleText: String? = nil
    if let title = map["title"] as? [String: Any] {
      titleType = title["type"] as? String
      titleText = title["text"] as? String
    }

    let centered = (map["alignment"] as? String) == "center"
    let coloredBackground = (map["coloredBackground"] as? Bool) ?? false

    return OctopusTopAppBarConfig(
      titleType: titleType,
      titleText: titleText,
      centered: centered,
      coloredBackground: coloredBackground
    )
  }

  // Parity wave — lifecycle: non-private so `switchCommunity` (added on
  // `OctopusReactNativeSdk`) can reuse the same decoding as `initialize`.
  func parseConnectionMode(from connectionModeMap: [String: Any], eventManager: OctopusEventManager) throws -> ConnectionMode {
    let connectionModeType = connectionModeMap["type"] as? String

    switch connectionModeType {
    case "sso":
      return try createSSOConnectionMode(from: connectionModeMap, eventManager: eventManager)
    case "octopus":
      // Parity wave — lifecycle: `deepLink` is a new field added in this PR — the JS
      // `connectionMode: { type: 'octopus' }` shape had no way to carry one before, so this
      // reads it now rather than restoring something that used to work.
      let deepLink = connectionModeMap["deepLink"] as? String
      return .octopus(deepLink: deepLink)
    default:
      throw InitializationError.invalidConnectionModeType
    }
  }

  private func createSSOConnectionMode(from connectionModeMap: [String: Any], eventManager: OctopusEventManager) throws -> ConnectionMode {
    let appManagedFields = ProfileFieldMapper.fromReactNativeArray(connectionModeMap["appManagedFields"] as? [String])

    return .sso(
      .init(
        appManagedFields: appManagedFields,
        loginRequired: {
          eventManager.emitLoginRequired()
        },
        modifyUser: { profileField in
          eventManager.emitEditUser(profileField: profileField)
        }
      )
    )
  }
}

enum InitializationError: Error, LocalizedError {
  case missingAPIKey
  case missingConnectionMode
  case invalidConnectionModeType

  var errorDescription: String? {
    switch self {
    case .missingAPIKey:
      return "API key is required"
    case .missingConnectionMode:
      return "Connection mode is required"
    case .invalidConnectionModeType:
      return "Invalid connection mode type"
    }
  }
}
